Add generic http configuration command line options

havocp opened the pull request.

havocp committed:
Remove bokeh_app and notebook props from CommandExecInfo

There's no reason these should be used as far as I know, and
we are really trying to contain knowledge of specific command
types to a limited area of the code.

havocp committed:
Add generic http server configuration options to commands

The idea here is that you can start an http server command
without special-casing Bokeh, Notebooks, or other kinds of command.

Also, bokeh and notebook commands will now both default to opening
a browser to the new http server (previously Bokeh did not).

`conda kapsel run` now supports options
--kapsel-host, --kapsel-port, --kapsel-url-prefix, and --kapsel-no-browser.
These are auto-translated into the appropriate options for Bokeh and notebooks,
and will be passed through unchanged from `conda kapsel run` to other commands.

A new command attribute `supports_http_options` is available,
which defaults to true for Bokeh and notebooks, and has to be set to true
manually for shell commands. For shell commands the http options are passed
through verbatim regardless of this attribute, but setting it to true
allows other code to ask whether a command can be run as an http server.

For bokeh and notebooks, setting it to false disables translation of
the --kapsel-* options, which probably isn't useful, but just being
complete. Maybe it's useful to disable if you don't want kapsel to
auto-add `--show` to a bokeh server or something.

havocp commented on conda_kapsel/project_commands.py:
```diff
@@ -28,6 +29,114 @@ def _is_windows():
+        raise RuntimeError("not implemented")  # pragma: no cover
+
+
+class _BokehArgsTransformer(_ArgsTransformer):
```
These "args transformers" could be factored out into some kind of loadable modules someday. Keeping it modular for now but not trying to make it actually extensible by third parties.

havocp commented on conda_kapsel/project_commands.py:
```diff
@@ -28,6 +29,114 @@ def _is_windows():
+_ArgSpec = namedtuple('_ArgSpec', ['option', 'has_value'])
+
+_http_specs = (_ArgSpec('--kapsel-host', True), _ArgSpec('--kapsel-port', True), _ArgSpec('--kapsel-url-prefix', True),
+               _ArgSpec('--kapsel-no-browser', False))
```
I used these namespaced command line options because I was afraid of stomping on existing options with our own that might have slightly different semantics or something. Since these options are intended mostly for use by scripts, it doesn't seem so bad. Also it helps explain why these options won't be in `--help`

havocp commented on conda_kapsel/project_commands.py:
```diff
@@ -28,6 +29,114 @@ def _is_windows():
+            if option in ('--kapsel-host', '--kapsel-port'):
+                for v in values:
+                    added.append(option.replace('kapsel-', ''))
+                    added.append(v)
```
I supported multiple copies of every option (such as --kapsel-host), but it only makes sense for some of them (such as host) and not for others (such as port). I figure let the invoked tool such as bokeh complain about that. Also I should add test coverage for the multiple instances case.

quasiben commented:
Looks good. Also,can you push something we could use a canonical example for this?

havocp committed:
Add a test for multiple --kapsel-host options passed in

havocp committed:
Add an example that uses Falcon to export a simple API

havocp committed:
Add --{no-}supports-http-options flags to add-command

havocp committed:
Attempt to fix exception on appveyor

The failure is

_______________________ test_conda_install_no_packages ________________________
[gw1] win32 -- Python 3.4.5 C:\Miniconda3-x64\envs\test-environment\python.exe
monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0x0000000006057198>
    def test_conda_install_no_packages(monkeypatch):
        monkeypatch_conda_not_to_use_links(monkeypatch)

        def do_test(dirname):
            envdir = os.path.join(dirname, "myenv")

            conda_api.create(prefix=envdir, pkgs=['python'])

            with pytest.raises(TypeError) as excinfo:
                conda_api.install(prefix=envdir, pkgs=[])
            assert 'must specify a list' in repr(excinfo.value)

>       with_directory_contents(dict(), do_test)
conda_kapsel\internal\test\test_conda_api.py:129:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
conda_kapsel\internal\test\tmpfile_utils.py:60: in with_directory_contents
    return func(os.path.realpath(dirname))
conda_kapsel\internal\test\test_conda_api.py:123: in do_test
    conda_api.create(prefix=envdir, pkgs=['python'])
conda_kapsel\internal\conda_api.py:110: in create
    return _call_conda(cmd_list)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
extra_args = ['create', '--yes', '--quiet', '--prefix', 'C:\\projects\\kapsel\\build\\tmp\\test-o81cvabt\\myenv', 'python']
    def _call_conda(extra_args):
        cmd_list = _get_conda_command(extra_args)

        try:
            p = subprocess.Popen(cmd_list, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        except OSError as e:
            raise CondaError("failed to run: %r: %r" % (" ".join(cmd_list), repr(e)))
        (out, err) = p.communicate()
        errstr = err.decode().strip()
        if p.returncode != 0:
>           raise CondaError('%s: %s' % (" ".join(cmd_list), errstr))
E           conda_kapsel.internal.conda_api.CondaError: conda create --copy --yes --quiet --prefix C:\projects\kapsel\build\tmp\test-o81cvabt\myenv python: An unexpected error has occurred.
E           Please consider posting the following information to the
E           conda GitHub issue tracker at:
E
E               https://github.com/conda/conda/issues
E
E
E
E           Current conda install:
E
E                          platform : win-64
E                     conda version : 4.2.7
E                  conda is private : False
E                 conda-env version : 4.2.7
E               conda-build version : not installed
E                    python version : 3.4.3.final.0
E                  requests version : 2.11.1
E                  root environment : C:\Miniconda3-x64  (writable)
E               default environment : C:\Miniconda3-x64\envs\test-environment
E                  envs directories : C:\Miniconda3-x64\envs
E                     package cache : C:\Miniconda3-x64\pkgs
E                      channel URLs : https://repo.continuum.io/pkgs/free/win-64/
E                                     https://repo.continuum.io/pkgs/free/noarch/
E                                     https://repo.continuum.io/pkgs/pro/win-64/
E                                     https://repo.continuum.io/pkgs/pro/noarch/
E                                     https://repo.continuum.io/pkgs/msys2/win-64/
E                                     https://repo.continuum.io/pkgs/msys2/noarch/
E                       config file : C:\Users\appveyor\.condarc
E                      offline mode : False
E
E
E
E           `$ C:\Miniconda3-x64\Scripts\conda-script.py create --copy --yes --quiet --prefix C:\projects\kapsel\build\tmp\test-o81cvabt\myenv python`
E
E
E
E
E               Traceback (most recent call last):
E                 File "C:\Miniconda3-x64\lib\site-packages\conda\exceptions.py", line 472, in conda_exception_handler
E                   return_value = func(*args, **kwargs)
E                 File "C:\Miniconda3-x64\lib\site-packages\conda\cli\main.py", line 144, in _main
E                   exit_code = args.func(args, p)
E                 File "C:\Miniconda3-x64\lib\site-packages\conda\cli\main_create.py", line 68, in execute
E                   install(args, parser, 'create')
E                 File "C:\Miniconda3-x64\lib\site-packages\conda\cli\install.py", line 405, in install
E                   execute_actions(actions, index, verbose=not context.quiet)
E                 File "C:\Miniconda3-x64\lib\site-packages\conda\plan.py", line 643, in execute_actions
E                   inst.execute_instructions(plan, index, verbose)
E                 File "C:\Miniconda3-x64\lib\site-packages\conda\instructions.py", line 134, in execute_instructions
E                   cmd(state, arg)
E                 File "C:\Miniconda3-x64\lib\site-packages\conda\instructions.py", line 58, in EXTRACT_CMD
E                   extract(arg)
E                 File "C:\Miniconda3-x64\lib\site-packages\conda\install.py", line 800, in extract
E                   with tarfile.open(fname) as t:
E                 File "C:\Miniconda3-x64\lib\tarfile.py", line 1548, in open
E                   return func(name, "r", fileobj, **kwargs)
E                 File "C:\Miniconda3-x64\lib\tarfile.py", line 1613, in gzopen
E                   fileobj = gzip.GzipFile(name, mode + "b", compresslevel, fileobj)
E                 File "C:\Miniconda3-x64\lib\gzip.py", line 181, in __init__
E                   fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
E               FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Miniconda3-x64\\pkgs\\python-3.5.2-0.tar.bz2'
conda_kapsel\internal\conda_api.py:51: CondaError
=================== 1 failed, 688 passed in 134.22 seconds ====================
pytest failed, code 1
pep257 says docstrings look good.

havocp committed:
Change appveyor workaround to lock conda 4.2.6 instead of 4.2.7

havocp committed:
Try conda 4.1.12 for appveyor build

havocp committed:
test_project.py: Add some more debugging around removing echo_stuff.bat

Same Windows problem we have been having...

havocp committed:
Even more debug spam about failure to remove echo_stuff.bat

havocp committed:
Add logging of os.stat output to test_project.py failing test
